lib/libc/gen/emutls.c

```diff
@@ -0,0 +1,311 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2019 Andrew Turner
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory under DARPA/AFRL contract FA8750-10-C-0237
+ * ("CTSRD"), as part of the DARPA CRASH research programme.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <sys/param.h>
+
+#include <machine/cpu.h>
+
+#include <pthread.h>
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef EMUTLS_LIBC
+#include "static_tls.h"
+#include "libc_private.h"
+#endif
+
+struct emutls_control {
+	uint64_t size;
+	uint64_t align;
+	union {
+		_Atomic size_t index;
+		void *unused; /* Ensure this is pointer sized */
+	};
+	void *value_addr;
+};
+
+_Static_assert(sizeof(struct emutls_control) == (sizeof(uint64_t) * 2 +
+    sizeof(void *) * 2), "emutls_control is the wrong size");
+
+struct emutls_array {
+	size_t size;
+	void *data[];
+};
+
+#ifdef EMUTLS_LIBC
```

Comment:
These `#ifdef`s seem pointless, they're always true?

Reply:
Hmmm, perhaps that is the case now.  It wasn't always set in the history of that file, but it would simplify the code a bit to drop the !EMUTLS_LIBC case.

Reply:
I don't see any reason not to drop the ifdefs. Building the file at all depends on the same conditional as defining EMUTLS_LIBC.

Reply:
The only question I have is is there any reason we might want to use this code outside of libc (e.g. would we potentially ever need to use it in libthr?).  If not, then I'm happy to drop them.  I think that's more of an @zxombie question as he added the #ifdef's originally.

Reply:
Well libc is what initialises TLS for statically-linked programs. The only thing you might possibly want is to link it into rtld. But the real answer is to get a stable TLS spec implemented in the Morello toolchain so we can ditch this and enter the 21st century.

Reply:
I would like that solution as well, but we will have to live with this until we have that solution. :)  Is dynamic TLS not currently working on Morello then or is it calling the libc functions?

Reply:
It supposedly works, we just don't know what the resolver interface actually is so we can't implement it in libc/rtld. I don't particularly want to continue trying to reverse-engineer it to guess what they did, and presumably the bionic implementation is only for static linking (and I wouldn't trust it to get everything right...).

Reply:
I meant today in CheriBSD does dynamic TLS work on morello using -femulated-tls (or is that a static-only thing?)

Reply:
Yeah, programs and libraries are using the dynamically-linked libc with -femulated-tls, and rtld doesn't use TLS anyway so things work fine.

Reply:
It was added so I could test the code without having to build it into a Morello libc.

Reply:
Do we think we still need to do that or is that no longer necessary?

Reply:
Andy confirms we won't need to debug this standalone anymore.